### ISSUE TITLE
while always returns null

### ISSUE THREAD

**Issue by czerwinskilukasz1** [xFAANG/askql#239]
`if` is essentially a `while` loop that is executed once at most, however return values for `if` and `while` are inconsistent.

If returns the value of its last statement as its value:
```
🦄 .editor
// Entering editor mode (^D to finish, ^C to cancel)
if(true) {
  let b = 4 
}

int ask(if(call(get('true')),block(let('b',4)),block()))
4
```

However while always returns null:
```
🦄 .editor
// Entering editor mode (^D to finish, ^C to cancel)
let b = 2
while(b<4) {
  b = b+1 
}

empty ask(let('b',2),while(call(get('<'),get('b'),4),block(assign('b',call(get('+'),get('b'),1)))))
null
🦄 
```

**Comment by mhagmajer** [comment]
@czerwinskilukasz1  What is the use case for `while` returning a value? You can encapsulate each imperative construct in AskScript with a function.

The case of `if` is a little different because we merged `if` and the ternary operator `?:` from JavaScript, allowing `if` to return a value.

**Comment by czerwinskilukasz1** [comment]
I'd say while should return its last result, and if not run at all, `empty`.
It's just a matter of being consistent - why would we return value from `if` but not from `while`?

**Comment by mhagmajer** [comment]
@czerwinskilukasz1 I think I explained about `if` in the previous message. 

Can you give some example programs that depend on `while` returning a value vs creating a function that does it? Do you know any other programming language that returns a value from a loop?

Interesting :)

**Comment by czerwinskilukasz1** [comment]
Nope, none, but I thought it's interesting. Aaaand, IIRC around a month ago `while` returned the value and I liked it.

**Comment by mhagmajer** [comment]
IIRC?

**Comment by mhagmajer** [comment]
I'm also interested in exploring this topic. `Promise.all` returns a value from a list of computations - kind of a loop

**Comment by czerwinskilukasz1** [comment]
If I Recall Correctly

**Comment by mhagmajer** [comment]
@czerwinskilukasz1 do you think it makes sense for `while` to return a collection of values from all iterations or would it just be a waste of memory?

**Comment by czerwinskilukasz1** [comment]
Oh, I was thinking about the _last_ value only.
I would say we shouldn't remember all the previous values.

**Comment by mhagmajer** [comment]
So you're thinking that `while` should be like`reduce`?

**Comment by mhagmajer** [comment]
@czerwinskilukasz1 is our parser going to handle that?
```
log(while(i < 100) { i = i + 1 })
```

**Comment by mhagmajer** [comment]
Same can be done with explicit recursion:

```
let i = 0
let f = fun() { if (!(i < 100)) { return }; i = i + 1; f() }
log(f())
```

So `while` is significantly shorter 👍

**Comment by czerwinskilukasz1** [comment]
Our parser _could_ handle it.  
Currently `if`, `while`, `for` etc. are _statements_, while function arguments could be _values_ (aka _expressions_) only. So I'd need to move statements there, *however* I think it opens a broad field for crappy code, loops inside arguments and stuff, so if I needed to choose all or nothing, I think I'd resign from while returning any value.

**Comment by czerwinskilukasz1** [comment]
Tasks:
  - [ ] add `while`, `for` (x3), `if` as a value in grammar 
  - [ ] return value from while and for (x3) in while/for....ts files
